### PR TITLE
Fixed #29147 -- Made JSONField.to_python() parse value as JSON.

### PR DESCRIPTION
**Pull request by kingbuzzman**
https://code.djangoproject.com/ticket/29147

### CODE REVIEW

**Comment by charettes** [comment]
@kingbuzzman it looks like the suggested changes are breaking backward compatiblity as Tim mentioned on the ticket. How you planning on addressing that?

On the ticket you've mentioned that `JSONField` should behave this way because `HStoreField` does. 

One fundamental difference between `HStoreFied` and `JSONField` is that the latter allow top level string values while the other doesn't. That makes `JSONField` special in how it has to deal with string because a string value cannot necessarily be considered as a serialized JSON structure. I'm not sure if we have extensive testing in this regards but I'd assume we'd need to perform some gymnastic like `django.forms.contrib.postgres.forms.JSONField` does to get everything right.

@sauravjaiswalsj out of curiosity, what makes you think this PR is ready to ready to get merged given it fails at least one test from the suite?

**Comment by kingbuzzman** [comment]
@charettes yup, i clearly see them not passing :( i cannot, for the life of me replicate this locally. I've tried different versions of postgres (9.3 flat out fails because of the lack of jsonb support) (9.4+ pass). Can you some insight into how i can replicate this locally?

this is how im running it:

`/django/tests/runtests.py postgres_tests.test_json.TestSerialization --failfast --settings=mysettings`

mysettings.py:

```
SECRET_KEY = 'hellooooo'
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'django',
        'USER': 'postgres',
        'PASSWORD': 'django',
        'HOST': '172.17.0.1',
        'PORT': 9876,
    }
}
```

everything passes over here :(

File: django/contrib/postgres/fields/jsonb.py [hunk 1]
```diff
@@ -41,6 +41,21 @@ def __init__(self, verbose_name=None, name=None, encoder=None, **kwargs):
         self.encoder = encoder
         super().__init__(verbose_name, name, **kwargs)
 
+    def to_python(self, value):
+        if isinstance(value, str):
+            try:
+                value = json.loads(value)
+            except json.decoder.JSONDecodeError:
+                try:
+                    value = json.loads('"' + value + '"')
```
Comment (kingbuzzman) [inline django/contrib/postgres/fields/jsonb.py]: This is beyond crap -- i'm open for suggestions...

**Comment by timgraham** [comment]
I believe the solution is ugly because your assumption about how `to_python()` is supposed to work is incorrect. Your use case and expectation about how `to_python()` should work is different than how the field works.

**Comment by kingbuzzman** [comment]
@timgraham according to the `to_python()` doc's, it says:

```
Convert the input value into the expected Python data type, raising
django.core.exceptions.ValidationError if the data can't be converted.
Return the converted value. Subclasses should override this.
```

https://github.com/django/django/blob/4b2bd30d8c67414ece8fc52f5d2027ace751d819/django/db/models/fields/__init__.py#L561

> your assumption about how to_python() is supposed to work is incorrect

I've since removed the `ValidationError` to better adjust to your regression requirement, i really do want to understand what im not understanding.

**Comment by akulakov** [comment]
@kingbuzzman Consider that I may store '{}' in a json field, that I may be expecting to be a string type that contains e.g. a placeholder for python string formatting (.format() method). Your code would convert it to an empty dict which would be unexpected value, as I'm reasonably expecting a string, and would also be backwards-incompatible, since code may be relying on getting a string in such case.

**Comment by timgraham** [comment]
I'll add the tests I developed for this branch in https://github.com/django/django/pull/9809.

Javier, I think the takeaway is that if your application makes this assumption about your data, you'll have to add your own parsing.

**Pull request closed without merging**